stdlib/public/TensorFlow/Gradients.swift

```diff
@@ -635,3 +635,67 @@ func _vjpRelu<T : TensorFlowFloatingPoint>(
 ) -> (Tensor<T>, (Tensor<T>) -> Tensor<T>) {
   return (relu(x), { v in Tensor(x .> 0) * v })
 }
+
+//===----------------------------------------------------------------------===//
+// Broadcasting
+//===----------------------------------------------------------------------===//
+
+extension Tensor where Scalar : TensorFlowFloatingPoint {
+  @inlinable
+  func _vjpBroadcast(
+    toShape shape: Tensor<Int32>
+  ) -> (Tensor, (Tensor) -> Tensor) {
+    return (broadcast(toShape: shape), { [origShape = self.shapeTensor] v in
+      v.unbroadcast(toShape: origShape)
+    })
+  }
+  
+  @inlinable
+  func _vjpBroadcast(
+    to shape: TensorShape
+  ) -> (Tensor, (Tensor) -> Tensor) {
+    return (broadcast(to: shape), { [origShape = self.shape] v in
+      v.unbroadcast(to: origShape)
+    })
+  }
+  
+  @inlinable
+  func _vjpBroadcast<OtherScalar>(
+    like other: Tensor<OtherScalar>
+  ) -> (Tensor, (Tensor) -> Tensor)
+    where OtherScalar : TensorFlowScalar {
+    return (broadcast(like: other), { [origShape = self.shapeTensor] v in
+      v.unbroadcast(like: origShape)
+    })
+  }
+}
+
+extension Tensor where Scalar : Numeric {
```

Comment:
Why is it "Scalar : Numeric" here but "Scalar : TensorFlowFloatingPoint" for _vjpBroadcast()?

Reply:
Right. This extension is incorrect. `Scalar` should be constrained on `TensorFloatFloatingPoint` because `Tensor` is only differentiable when `Scalar` conforms to `TensorFlowFloatingPoint`.

Reply:
The unbroadcast functions are defined within 
```swift
public extension Tensor where Scalar : Numeric
```
while the broadcast functions are defined within:
```swift
public extension Tensor
```

so it seems that `Scalar` has to be `Numeric` for the unbroadcast functions, which is why I added that constraint.

https://github.com/apple/swift/blob/tensorflow/stdlib/public/TensorFlow/Ops.swift#L1620

Reply:
However, I see further constraining unbroadcast to be `@differentiable(wrt: self, vjp: _vjpUnbroadcast(toShape:) where Scalar : TensorFlowFloatingPoint)` fixes this